File: LeetCode/Easy/0496_Next_Greater_Element_1/Solution.java
// 2022-09-17 (stack)
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer, Integer> nextLargest = new HashMap<>();
        Deque<Integer> stack = new ArrayDeque<Integer>(); // only holds values from high to low (at top)
        
        for (int num : nums2) {
            // while the current value is greater than the top of our stack, record down as next largest and pop
            while (stack.size() > 0 && stack.peek() < num) {
                nextLargest.put(stack.peek(), num);
                stack.pop();
            }
            stack.push(num);
        }
        
        int[] nextGreaters = new int[nums1.length];
        Arrays.fill(nextGreaters, -1);
        for (int i=0; i<nums1.length; i++) {
            if (nextLargest.containsKey(nums1[i])) {
                nextGreaters[i] = nextLargest.get(nums1[i]);
            }
        }
        return nextGreaters;
    }
}

File: LeetCode/Easy/0496_Next_Greater_Element_1/Solution.py
## 2022-01-04 (stack)
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        """
            Using a stack, iterate through nums2 to determine the next
            greatest element. We only store values in the stack from greatest (bottom)
            to lowest (top). 
            
            If we encounter a number that is greater than the top of
            our stack, we assign the top number's next greatest as the number we've 
            encountered in a dictionary, and pop it off our stack. 
            
            We continue until the top of the stack is greater than our current number, or 
            if our stack runs out. We then push the current number onto our stack and continue
            until we finish iterating through nums2.
            
            We can then generate the output list using our dictionary.

            This works since there's a guarantee that all numbers in nums1 and nums2 will
            be unique, so our greater assignments won't overwrite each other.
        """
        greater = dict()
        stack = []
        for num in nums2:
            while stack and num > stack[-1]:
                greater[stack[-1]] = num
                stack.pop()
            stack.append(num)
        return [greater[num] if num in greater else -1 for num in nums1]